app/code/core/Mage/Usa/Model/Shipping/Carrier/Ups.php

```diff
@@ -1201,62 +1205,62 @@ protected function _parseRestTrackingResponse($trackingValue, $jsonResponse)
         if ($jsonResponse) {
             $responseData = json_decode($jsonResponse, true);
 
-            if ($responseData['trackResponse']['shipment']) {
+            if (isset($responseData['trackResponse']['shipment'][0]['package'][0]['activity'])) {
```

Comment:
I'm not sure I agree with this change. If there isn't tracking info, that's not necessarily an error, right? If we give a tracking number that UPS doesn't have tracking for anymore (they only track for 120 days), do we want to display the error message: "For some reason we can\'t retrieve tracking info right now."? That's not quite right, IMO. 

I think the way you had it before was probably better. It's not wrong to return an empty array from this function if the request was successful, but didn't have any "activity".

UPS will return "{"trackResponse":{"shipment":[{"warnings":[{"code":"TW0001","message":"Tracking Information Not Found"}]}]}}" for such a case.

Reply:
exactly, that's why I did it, becuase in the elseif I'm catching the "warning", but let me check better

Reply:
i don't get it, if there's no activity the whole "if" won't do anything

Reply:
It's actually okay for the whole "if" to do nothing. If you follow the call chain, it eventually goes up to `Mage_Sales_Model_Order_Shipment_Track::getNumberDetail`, line 147, where it handles an empty tracking result just fine.

Reply:
yes but it will generated warning for non-existent  array keys I think

Reply:
since the whole if doesn't do anything if there's no activity, I think it's better to protect it with a full isset like now, I could remove the first elseif but... why? if there's a specific warning message we will return it

Reply:
> yes but it will generated warning for non-existent array keys I think

It won't. The `??` operator prevents PHP from complaining about it.

> if there's a specific warning message we will return it

The only reason I'd hesitate here is because I don't know every warning that UPS could send back. What if it sends something that is very "programmer-y" that only the site administrators should be seeing? End users won't want to see errors about missing request parameters or something. 

Maybe just logging warnings would make more sense?

Reply:
I see the point of weird messages from UPS but "tracking not found" I think it's pretty important.

Btw to me it's not finding any tracking :-\

Reply:
> I see the point of weird messages from UPS but "tracking not found" I think it's pretty important.

That one specifically *isn't* super important because if this function sets an empty trackingResult, then the calling classes handle it by displaying: `Mage::helper('sales')->__('No detail for number "%s"', $this->getNumber());`. So that case is handled "automatically" for us.


> Btw to me it's not finding any tracking :-\

Do you have a specific tracking number I can test? Is the order recent?

Reply:
I've reverted it, but I don't like it that much.

About tracking, sorry, it works, it's just weird that you can click "track order" also if there's no tracking number, but that's another something